### PR TITLE
Docker: disable compsec (distorts benchmarks)

### PR DESCRIPTION
**Pull request by olliemath**
## Description

Disables seccomp when running docker.

For unknown reasons, running python under docker can be over 50% slower than native usage. This is a performance bug in docker, which has something to do with its use of seccomp. See https://pythonspeed.com/articles/docker-performance-overhead/
and the corresponding docker issue https://github.com/moby/moby/issues/41389

Practically, the performance difference is clear - locally when running `make DIRECTORY=PrimePython`, on a linux system the performance of `ssovtest`'s solution goes from **1998** passes to **4110**. It seems seccomp doesn't affect the usage of compiled languages (in particular the non-faithful numpy solution seems unaffected), but you *may* see a speed boost for some other interpreted languages with this fix.

### Other considerations

Should we disable seccomp if it's providing security? Well, if all code in this repo has been reviewed and verified, it should be absolutely fine (note all the containers are running as root user anyway, so we better hope this is the case).

## Contributing requirements

* [x] I read the contribution guidelines in CONTRIBUTING.md.
* [x] I placed my solution in the correct solution folder.
* [ ] I added a README.md with the right badge(s).
* [ ] I added a Dockerfile that builds and runs my solution.
* [x] I selected `drag-race` as the target branch.
* [x] All code herein is licensed compatible with BSD-3.

~

### CODE REVIEW

**Comment by rbergen** [comment]
@marghidanu I think you're better suited to review this PR than I am, so I'll leave it to you in first instance

**Comment by marghidanu** [comment]
I would make this feature optional, to give the user the option to disable it on demand.

**Comment by olliemath** [comment]
OK, I've added an `UNCONFINED` option, which I've set to `true` by default.

If you'd like it to be `false` by default, revert the final commit.

File: BENCHMARK.md [hunk 1]
```diff
@@ -335,9 +335,20 @@ Make the `DIRECTORY` variable point to the directory that contains the solutions
 make DIRECTORY=PrimeCPP
 ```
 
+## Running with seccomp sandboxing
+
+For some interpreted languages (Python, Ruby, NodeJS), docker has a non-zero affect slowing on CPU-intensive code
```
Comment (tjol) [inline BENCHMARK.md]: `s/affect slowing on/effect slowing down/`?

Definitely effect, not affect.

**Comment by rbergen** [comment]
> If you'd like it to be `false` by default, revert the final commit.

Please do make `false` the default. In line with the security by default principle, seccomp should be _enabled_ unless a user chooses to _disable_ it through deliberate action.

File: Makefile [hunk 2]
```diff
@@ -4,14 +4,15 @@ SHELL := /bin/bash
...
+UNCONFINED := "true"
 
 .PHONY: all
 all: benchmark
 
 .PHONY: benchmark
 benchmark: check-env
 	@REALPATH=$$(cd "$${DIRECTORY}" && pwd); \
-	ARGS=("-d $${REALPATH}" "-f $(FORMATTER)"); \
+	ARGS=("-d $${REALPATH}" "-f $(FORMATTER)" "-u $(UNCONFINED)"); \
```
Comment (marghidanu) [inline Makefile]: Please make this a flag without having to specify true/false values.

File: tools/src/commands/benchmark.ts [hunk 3]
```diff
@@ -29,8 +29,12 @@ export const command = new Command('benchmark')
   .requiredOption('-d, --directory <directory>', 'Implementation directory')
   .option('-f, --formatter <type>', 'Output formatter', 'table')
   .option('-o, --output-file <file>', 'Write output to given file')
+  .option('-u, --unconfined <bool>', 'Run with seccomp:unconfined (native performance for interpreted languages)')
   .action(async (args) => {
     const directory = path.resolve(args.directory as string);
+    const unconfined = args.unconfined === 'true';
```
Comment (marghidanu) [inline tools/src/commands/benchmark.ts]: Just the presence of the flag should be enough.

File: tools/src/services/docker.ts [hunk 4]
```diff
@@ -7,8 +7,13 @@ export default class DockerService {
     });
   }
 
-  public runContainer(imageName: string): string {
-    const output = child_process.execSync(`docker run --rm ${imageName}`, {
+  public runContainer(imageName: string, unconfined: boolean): string {
```
Comment (marghidanu) [inline tools/src/services/docker.ts]: Rather than having `unconfined` as a single parameter please add a list of options.

File: Makefile (hunk 2 above)
Comment (olliemath) [reply Makefile]: I'm afraid I don't know enough makefile voodoo for that - my googling is not sufficient to reveal how to do
```
make something --flag
```

Comment (marghidanu) [reply Makefile]: There is no need for voodoo; please look at the OUTPUT_FILE variable one line below and how it's set conditionally. ;)

Comment (olliemath) [reply Makefile]: OK, I've done it that way - it's now enabled only if `UNCONFINED` is set. Of course, doing it this way means that
```
make UNCONFINED=false
```
will also turn **on** unconfined mode

Comment (marghidanu) [reply Makefile]: It just needs to be set, that’s fine

**Comment by marghidanu** [comment]
LGTM

**Review by marghidanu — APPROVED** [review]

**Pull request merged**